# RARE-Seq Analysis Workflow

The RARE-Seq analysis workflow consists of the following steps:

1. Import RSEM expected counts. This step uses the tximport R package to merge expected counts, expected lengths, and transcript per million (TPM) values generated by RSEM into a single R object to be used by downstream steps.
2. Normalize expression. This step calculates log-transformed and normalized gene expression (log2NX) for captured coding genes across all samples. Expression normalization may be done using TPM (from RSEM) or TMM (Trimmed Mean of M values method from edgeR). For the TMM method, a subset of samples may be labeled as “meta-reference” in the sample file, which will be used to calculate the gene-wise M values. In addition, platelet-based correction may be done using RUVg or RUVs methods.

For cfRNA enrichment score (ES) analysis:

3. Calculate z-scores. This step calculates gene-wise z-scores for all samples. Z-scores are calculated using the average and standard deviation of gene expression in the subset of samples that have been labeled as “meta-reference” in the sample file input. If no meta-reference is provided, the entire cohort is used as the meta-reference.
4. Calculate enrichment scores. This step calculates sample-wise enrichment scores for the input gene signature(s). Gene signatures are filtered to include only the genes that intersect with the Z-score matrix input into this step.

For differential expression analysis:

5. Identify differentially expressed genes. This step uses the DESeq2 R package to determine differentially expressed genes according to the specified model equation.

For building cancer versus control elastic net classifier:

6. Set up elastic net model training inputs. This step combines gene expression and variant-based features into a single matrix to be used for the elastic net classifier model training. Features are then filtered to include 1) genes that are expressed in either LUAD tumor tissue or meta-reference cfRNA, 2) genes that are significantly differentially expressed when comparing LUAD tumor tissue to meta-reference cfRNA, and 3) genes that are not associated with sex (i.e., genes that are significantly differentially expressed in male versus female meta-reference cfRNA, or specific to reproductive cell types in PanglaoDb, or on sex chromosomes). Pre-processing is done to remove features with near-zero variance in the training cohort, followed by standardization. In addition, samples are split into training and validation sets, and the training set is further separated into N folds to be used for model cross validation.
7. Train elastic net model. This step uses the caret and glmnet R packages to train an elastic net model with the selected samples and selected features. Model training uses feature weights for feature-specific regularization. Model training is further restricted so that only features with positive coefficients are selected.
8. Validate elastic net model. This step uses the caret and glmnet R packages to predict cancer likelihood scores in the withheld validation samples.

The RARE-Seq analysis workflow has an expected runtime of about 5 minutes.

# Set up

Anaconda should be installed on linux operating system before running the RARE-Seq analysis workflow. Once conda is installed and initialized, the RARE-Seq analysis conda environment can be created. Example set-up commands are below.
```
conda init bash
conda update -n base conda
conda config --add channels defaults
conda config --add channels bioconda
conda config --add channels conda-forge
conda config --set channel_priority false
conda env create -f rareseq_env.yml -n rareseq_env
```

The environment contains the following dependencies:

- bioconductor-biocparallel=1.28.3
- bioconductor-deseq2=1.34.0
- bioconductor-edger=3.36.0
- bioconductor-fgsea=1.20.0
- bioconductor-limma=3.50.3
- bioconductor-ruvseq=1.28.0
- bioconductor-tximport=1.22.0
- r-ashr=2.2_54
- r-base=4.1.3
- r-caret=6.0_93
- r-combinat=0.0_8
- r-data.table=1.14.2
- r-glmnet=4.1_2
- r-ggpubr=0.4.0
- r-mass=7.3_58.1
- r-mixtools=1.2.0
- r-proc=1.18.0
- r-tidyverse=1.3.2

The set-up steps above only need to be run once. On the other hand, the conda environment will need to be activated before each instance of running the RARE-Seq analysis workflow.

```
conda activate rareseq_env
```

## Step #1: Import RSEM expected counts.


### Example Usage:

```
Rscript importExpression.R <INPUT DIRECTORY> <SAMPLE FILE>
```

### Input Parameters:

1. *Input directory*: Path to single directory containing all RSEM output files with “.genes.results” suffix.
2. *Sample file*: A tab-delimited file listing samples to use for RARE-Seq analysis.
	a. The file must minimally include the column(s) below.
		 i. “ID”: Unique sample identifier
	b. Additional columns may be included.

### Output Files:

*Tximport RDS file*: An RDS file named “tximport.rds” containing data output by RSEM and re-formatted into a list of 3 matrices. The list contains 1) matrix with expected counts (“counts”), 2) matrix with expected length (“length”), and 3) matrix with transcripts per million (TPM) (“abundance”). For each matrix, genes are listed as rows and samples as columns.

## Step #2: Normalize expression.

### Example Usage:

```
Rscript normalizeExpression.R <TXIMPORT RDS FILE> <GENE FILE> <SAMPLE FILE> <NORMALIZATION METHOD> <CORRECTION METHOD> <NEGATIVE CONTROL GENE FILE>
```

### Input Parameters:

1. *Tximport RDS file*: The tximport RDS file output from step #1.
2. *Gene file*: A tab-delimited file containing the list of gene identifiers to include in the normalization and downstream steps.
	a. The file must minimally include the column(s) below.
		 i. “gene_id_simplified”: Ensembl gene identifier. The gene version number should be removed, e.g. “ENSG00000196230” instead of “ENSG00000196230.12”.
3. *Sample file*: A tab-delimited file listing samples to use for RARE-Seq analysis. All samples included in this file must have also been included in the sample file used for step #1, and the same sample file can be used for both steps.
	a. The file must minimally include the column(s) below.
		 i. “ID”: Unique sample identifier
		 ii. “group”: Relevant sample categorization (i.e., “cancer” and “control”)
		 iii. “Reference”: Meta-reference samples may be marked “yes” or “NA” to indicate whether the sample is or is not included in the meta-reference subset, respectively.
	b. The file may include the columns below if step #7 is to be run.
		 i. “EN: Sample labels including “train”, “validation”, or “NA”, if they are to be used for model training, used for model validation, or excluded (i.e., meta-reference samples), respectively.
	c. Additional columns may be included.
4. *Normalization method*: A character string for the appropriate normalization method and must be set to “NONE”, “TMM”, or “TPM”.
5. *Correction method*: A character string for the appropriate correction method and must be set to “NONE”, “RUVg”, or “RUVs”.
6. *Negative control gene file*: A tab-delimited file listing the genes to use for RUVg correction, and therefore is required if “RUVg” is specified above. For cfRNA analysis, the default is to use platelet marker genes for RUVg correction.
	a. The file must minimally include the column(s) below.
		 i. “gene_id_simplified”: Ensembl gene identifier. The gene version number should be removed, e.g. “ENSG00000196230” instead of “ENSG00000196230.12”.

### Output Files:

*Normalized RDS file*: An RDS file named according to the <NORMALIZATION METHOD> and <CORRECTION METHOD>” input. The list contains 1) data.frame with input gene information (“genes”), 2) data.frame with input sample information (“samples”), 3) matrix of RSEM expected counts (“counts”), 4) matrix of RSEM expected length (“length”), and 5) matrix of normalized gene expression (“log2NX”). For last 3 matrices, genes are listed as rows and samples as columns.

## Step #3: Calculate z-scores.

### Example Usage:

```
Rscript calcZcores.R <NORMALIZED RDS FILE>
```

### Input Parameters:

*Normalized RDS file*: The RDS file output from step #2.

### Output Files:

*Normalized RDS file with Z-scores*: An RDS file containing the same matrices as output for step #2 but with the matrix of z-scores added and with “_withZscores” appended to the file name. If meta-reference samples were specified in the sample information file in step #2, these samples will be used for Z-score standardization and will therefore be removed from the Z-score matrix.

## Step #4: Calculate enrichment scores.

### Example Usage:

```
Rscript calcEnrichment.R <NORMALIZED RDS FILE WITH ZSCORES> <ENRICHMENT METHOD> <GENE SIGNATURE FILE> <OUTPUT DIRECTORY>
```

### Input Parameters:

1. *Normalized RDS file with Z-scores*: The RDS file output from step #3.
2. *Enrichment method*: A character string specifying the desired method for calculating enrichment scores. The default method is “stouffer”, which uses Stouffer’s method to combine multiple weighted Z-scores into a single gene signature enrichment score. The method may also be “percentile_X” and if so, X should be replaced by a numeric value in [0,1] (e.g., 0.9) to be used as the cut point among weighted Z-scores within the gene signature defining the gene signature enrichment score. Finally, the method may be set to “sum”, which simply adds together the weighted Z-scores to calculate the gene signature enrichment score.
3. *Gene signature(s) file*: A tab-delimited file containing the list of genes in the gene signature(s) of interest.
a. The file must minimally include the column(s) below.
 i. “gene_id_simplified”: Ensembl gene identifier. The gene version number should be removed, e.g. “ENSG00000196230” instead of “ENSG00000196230.12”.
b. The file may include the columns below.
 i. “gene_set_name”: Gene set name(s). Multiple gene signatures may be run but must be distinguished with unique names in this column.
 ii. “weight”: Gene weight(s) to use for the specified enrichment method. If no weights are provided, all gene weights are set to 1.
4. *Output directory*: A character string providing the name of the directory in which to output the enrichment score results. Since this step can be run multiple times for multiple gene signature files, each run should use a unique output directory name to avoid overwriting data.

### Output Files:

*Enrichment scores file*: A tab-delimited file containing the enrichment scores and associated empirical p-values for each sample for each gene signature.

## Step #5: Identify differentially expressed genes.

### Example Usage:

```
Rscript differentialExpression.R <NORMALIZED RDS FILE> <DIFFERENTIAL EXPRESSION METHOD> <MODEL EQUATION>
```

### Input Parameters:

1. *Normalized RDS file*: The RDS file output from step #2.
2. *Differential expression method*: A character string specifying which differential expression method to use and must be either “DESeq2” or “DESeq2shrink”. The latter runs log fold change (LFC) shrinkage using the ashr method to provide better gene ranking by effect size. More information is provided in the DESeq2 tutorial and vignettes (https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html).
3. *Model equation*: A character string providing the equation used to model differential expression. This equation must use column headers that are present in the sample information matrix in the RDS file input (which come from the sample information file input in step #2). For example, for a simple case-control analysis, the model equation may be “~group” to determine differentially expressed genes between two (or more) sample groups as distinguished in this column. More information is provided in the DESeq2 tutorial and vignettes (https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html).

### Output Files:

*Differential expression file(s)*: A tab-delimited file containing the log fold change and adjusted p-values for differential expression based on the provided model equation. One file will be output for each contrast evaluated.

## Step #6: Set up elastic net training inputs.

### Example Usage:

```
Rscript setupEN.R <NORMALIZED RDS FILE> <GENOTYPING FILE>
```

### Input Parameters:

1. *Normalized RDS file*: The RDS file output from step #2. The sample info matrix must include a column using the header “EN” that specifies the subset of samples to be used for model training (“train”), the subset of samples to be used for model validation (“validation”), and meta-reference to be excluded from this step (“NA”).
2. *Genotyping file (Optional)*: The tab-delimited file output by the targeted-caller script, named “variant_call_table.txt”. If not provided, only gene expression features will be used for model training.

### Output Files:

*Normalized RDS file for Caret*: An RDS file with “_forCaret.rds” appended to the filename. The RDS object contains 1) data.frame of gene information that has been filtered to include only features to be used for model training, 2) data.frame of sample information that has been filtered to include only samples to be used for model training, 3) matrices of standardized features to be used as predictors for model training and validation, 4) binarized vectors to be used as outcome variable (e.g. case/1 or control/0) for model training and validation. See the Description section for more information about how feature filtering and pre-processing.

## Step #7: Train elastic net model.

### Example Usage:

```
Rscript trainEN.R <NORMALIZED RDS FILE FOR CARET>
```

### Input Parameters:

*Normalized RDS file for Caret*: The RDS file output from step #7.

### Output Files:

1. *Model selected features file*: A tab-delimited file named “model_selected_features.txt” that contains feature coefficients. A feature coefficient above zero indicates feature selection.
2. *Model sample predictions file*: A tab-delimited file named “model_sample_predictions.txt” that contains the cancer likelihood score for all training samples.
3. *Model metrics file*: A tab-delimited file named “model_metrics.txt” that contains area under the curve (AUC) calculated either using the full training set (“Train”) or using the 10-fold cross validation results (“CV”).
4. *Model RDS file*: An RDS file named “model.rds” that contains the final model as output by the caret R package.

## Step #8: Validate elastic net model.

### Example Usage:

```
Rscript validateEN.R <NORMALIZED RDS FILE FOR CARET> <MODEL RDS FILE>
```

### Input Parameters:

1. *Normalized RDS file for Caret*: The RDS file output from step #7.
2. *Model RDS file*: The RDS file output from step #8.

### Output Files:

1. *Model sample predictions file*: A tab-delimited file named “model_sample_predictions_validation.txt” that contains the cancer likelihood score for all withheld validation samples.
2. *Model metrics file*: A tab-delimited file named “model_metrics_validation.txt” that contains the area under the curve *AC) calculated using the validation set (“Validation”).
