<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html>
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
  <meta http-equiv="Content-Style-Type" content="text/css">
  <title></title>
  <meta name="Generator" content="Cocoa HTML Writer">
  <meta name="CocoaVersion" content="2487.7">
  <style type="text/css">
    p.p1 {margin: 0.0px 0.0px 10.7px 0.0px; font: 36.0px Arial; -webkit-text-stroke: #000000}
    p.p2 {margin: 0.0px 0.0px 10.7px 0.0px; font: 14.7px Arial; -webkit-text-stroke: #000000}
    p.p3 {margin: 0.0px 0.0px 0.0px 48.0px; text-indent: -48.0px; font: 14.7px Arial; -webkit-text-stroke: #000000}
    p.p4 {margin: 0.0px 0.0px 10.7px 48.0px; text-indent: -48.0px; font: 14.7px Arial; -webkit-text-stroke: #000000}
    p.p5 {margin: 0.0px 0.0px 0.0px 0.0px; font: 14.7px 'Courier New'; -webkit-text-stroke: #000000}
    p.p6 {margin: 0.0px 0.0px 10.7px 0.0px; font: 14.7px 'Courier New'; -webkit-text-stroke: #000000}
    p.p7 {margin: 0.0px 0.0px 0.0px 0.0px; font: 14.7px Arial; -webkit-text-stroke: #000000; min-height: 16.0px}
    p.p8 {margin: 0.0px 0.0px 0.0px 0.0px; font: 14.7px Arial; -webkit-text-stroke: #000000}
    p.p9 {margin: 0.0px 0.0px 0.0px 96.0px; text-indent: -96.0px; font: 14.7px Arial; -webkit-text-stroke: #000000}
    p.p10 {margin: 0.0px 0.0px 0.0px 144.0px; text-indent: -144.0px; font: 14.7px Arial; -webkit-text-stroke: #000000}
    p.p11 {margin: 0.0px 0.0px 10.7px 144.0px; text-indent: -144.0px; font: 14.7px Arial; -webkit-text-stroke: #000000}
    span.s1 {font-kerning: none}
    span.s2 {text-decoration: underline ; font-kerning: none}
    span.s3 {font: 9.3px 'Times New Roman'; font-kerning: none}
    span.s4 {font: 14.7px Arial; font-kerning: none}
    span.s5 {-webkit-text-stroke: 0px #000000}
    span.Apple-tab-span {white-space:pre}
  </style>
</head>
<body>
<p class="p1"><span class="s1">RARE-Seq Analysis Workflow</span></p>
<p class="p2"><span class="s1">Version 4.0</span></p>
<p class="p2"><span class="s2">Description</span></p>
<p class="p2"><span class="s1">The RARE-Seq analysis workflow consists of the following steps:</span></p>
<p class="p3"><span class="s1">1.</span><span class="s3">     </span><span class="s1"><b>Import RSEM expected counts</b>. This step uses the tximport R package to merge expected counts, expected lengths, and transcript per million (TPM) values generated by RSEM into a single R object to be used by downstream steps.</span></p>
<p class="p4"><span class="s1">2.</span><span class="s3">     </span><span class="s1"><b>Normalize expression</b>. This step calculates log-transformed and normalized gene expression (log2NX) for captured coding genes across all samples. Expression normalization may be done using TPM (from RSEM) or TMM (Trimmed Mean of M values method from edgeR). For the TMM method, a subset of samples may be labeled as “meta-reference” in the sample file, which will be used to calculate the gene-wise M values.  In addition, platelet-based correction may be done using RUVg or RUVs methods.</span></p>
<p class="p2"><span class="s1"><b><i>For cfRNA enrichment score (ES) analysis</i></b>:</span></p>
<p class="p3"><span class="s1">3.</span><span class="s3">     </span><span class="s1"><b>Calculate z-scores</b>. This step calculates gene-wise z-scores for all samples. Z-scores are calculated using the average and standard deviation of gene expression in the subset of samples that have been labeled as “meta-reference” in the sample file input. If no meta-reference is provided, the entire cohort is used as the meta-reference.</span></p>
<p class="p4"><span class="s1">4.</span><span class="s3">     </span><span class="s1"><b>Calculate enrichment scores</b>. This step calculates sample-wise enrichment scores for the input gene signature(s). Gene signatures are filtered to include only the genes that intersect with the Z-score matrix input into this step.</span></p>
<p class="p2"><span class="s1"><b><i>For differential expression analysis</i></b>:</span></p>
<p class="p4"><span class="s1">5.</span><span class="s3">     </span><span class="s1"><b>Identify differentially expressed genes</b>. This step uses the DESeq2 R package to determine differentially expressed genes according to the specified model equation.</span></p>
<p class="p2"><span class="s1"><b><i>For building cancer versus control elastic net classifier</i></b>:</span></p>
<p class="p3"><span class="s1">6.</span><span class="s3">     </span><span class="s1"><b>Set up elastic net model training inputs.</b> This step combines gene expression and variant-based features into a single matrix to be used for the elastic net classifier model training. Features are then filtered to include 1) genes that are expressed in either LUAD tumor tissue or meta-reference cfRNA, 2) genes that are significantly differentially expressed when comparing LUAD tumor tissue to meta-reference cfRNA, and 3) genes that are not associated with sex (i.e., genes that are significantly differentially expressed in male versus female meta-reference cfRNA, or specific to reproductive cell types in PanglaoDb, or on sex chromosomes). Pre-processing is done to remove features with near-zero variance in the training cohort, followed by standardization. In addition, samples are split into training and validation sets, and the training set is further separated into N folds to be used for model cross validation.</span></p>
<p class="p3"><span class="s1">7.</span><span class="s3">     </span><span class="s1"><b>Train elastic net model</b>. This step uses the caret and glmnet R packages to train an elastic net model with the selected samples and selected features. Model training uses feature weights for feature-specific regularization. Model training is further restricted so that only features with positive coefficients are selected.</span></p>
<p class="p4"><span class="s1">8.</span><span class="s3">     </span><span class="s1"><b>Validate elastic net model.</b> This step uses the caret and glmnet R packages to predict cancer likelihood scores in the withheld validation samples.</span></p>
<p class="p2"><span class="s1">The RARE-Seq analysis workflow has an expected runtime of about 5 minutes.<span class="Apple-converted-space"> </span></span></p>
<p class="p2"><span class="s2">Set up</span></p>
<p class="p2"><span class="s1">Anaconda should be installed on linux operating system before running the RARE-Seq analysis workflow. Once conda is installed and initialized, the RARE-Seq analysis conda environment can be created. Example set-up commands are below.</span></p>
<p class="p5"><span class="s1">conda init bash</span></p>
<p class="p5"><span class="s1">conda update -n base conda</span></p>
<p class="p5"><span class="s1">conda config --add channels defaults</span></p>
<p class="p5"><span class="s1">conda config --add channels bioconda</span></p>
<p class="p5"><span class="s1">conda config --add channels conda-forge</span></p>
<p class="p5"><span class="s1">conda config --set channel_priority false</span></p>
<p class="p5"><span class="s1">conda env create -f rareseq_env.yml</span><span class="s4"> </span><span class="s1">-n rareseq_env</span></p>
<p class="p2"><span class="s1"> </span></p>
<p class="p2"><span class="s1">The environment contains the following dependencies:</span></p>
<p class="p3"><span class="s1">-</span><span class="s3">       </span><span class="s1">bioconductor-biocparallel=1.28.3</span></p>
<p class="p3"><span class="s1">-</span><span class="s3">       </span><span class="s1">bioconductor-deseq2=1.34.0</span></p>
<p class="p3"><span class="s1">-</span><span class="s3">       </span><span class="s1">bioconductor-edger=3.36.0</span></p>
<p class="p3"><span class="s1">-</span><span class="s3">       </span><span class="s1">bioconductor-fgsea=1.20.0</span></p>
<p class="p3"><span class="s1">-</span><span class="s3">       </span><span class="s1">bioconductor-limma=3.50.3</span></p>
<p class="p3"><span class="s1">-</span><span class="s3">       </span><span class="s1">bioconductor-ruvseq=1.28.0</span></p>
<p class="p3"><span class="s1">-</span><span class="s3">       </span><span class="s1">bioconductor-tximport=1.22.0</span></p>
<p class="p3"><span class="s1">-</span><span class="s3">       </span><span class="s1">r-ashr=2.2_54</span></p>
<p class="p3"><span class="s1">-</span><span class="s3">       </span><span class="s1">r-base=4.1.3</span></p>
<p class="p3"><span class="s1">-</span><span class="s3">       </span><span class="s1">r-caret=6.0_93</span></p>
<p class="p3"><span class="s1">-</span><span class="s3">       </span><span class="s1">r-combinat=0.0_8</span></p>
<p class="p3"><span class="s1">-</span><span class="s3">       </span><span class="s1">r-data.table=1.14.2</span></p>
<p class="p3"><span class="s1">-</span><span class="s3">       </span><span class="s1">r-glmnet=4.1_2</span></p>
<p class="p3"><span class="s1">-</span><span class="s3">       </span><span class="s1">r-ggpubr=0.4.0</span></p>
<p class="p3"><span class="s1">-</span><span class="s3">       </span><span class="s1">r-mass=7.3_58.1</span></p>
<p class="p3"><span class="s1">-</span><span class="s3">       </span><span class="s1">r-mixtools=1.2.0</span></p>
<p class="p3"><span class="s1">-</span><span class="s3">       </span><span class="s1">r-proc=1.18.0</span></p>
<p class="p4"><span class="s1">-</span><span class="s3">       </span><span class="s1">r-tidyverse=1.3.2</span></p>
<p class="p2"><span class="s1"> </span></p>
<p class="p2"><span class="s1">The set-up steps above only need to be run once. On the other hand, the conda environment will need to be activated before each instance of running the RARE-Seq analysis workflow.</span></p>
<p class="p6"><span class="s1">conda activate rareseq_env</span></p>
<p class="p2"><span class="s1"> </span></p>
<p class="p7"><span class="s1"></span><br></p>
<p class="p2"><span class="s2"> </span></p>
<p class="p2"><span class="s2">Step #1:</span><span class="s1"> Import RSEM expected counts.</span></p>
<p class="p2"><span class="s1">Example Usage:</span></p>
<p class="p6"><span class="s1">Rscript importExpression.R &lt;INPUT DIRECTORY&gt; &lt;SAMPLE FILE&gt;</span></p>
<p class="p2"><span class="s1">Input Parameters:</span></p>
<p class="p3"><span class="s1">1.</span><span class="s3">     </span><span class="s1"><b>Input directory</b> Path to single directory containing all RSEM output files with “.genes.results” suffix.</span></p>
<p class="p3"><span class="s1">2.</span><span class="s3">     </span><span class="s1"><b>Sample file</b> A tab-delimited file listing samples to use for RARE-Seq analysis.</span></p>
<p class="p3"><span class="s1"><span class="Apple-tab-span">	</span>a. The file must minimally include the column(s) below.</span></p>
<p class="p8"><span class="s3"><span class="Apple-tab-span">	</span><span class="Apple-tab-span">	</span> </span><span class="s1">i.</span><span class="s3">     </span><span class="s1">“ID”: Unique sample identifier</span></p>
<p class="p8"><span class="s3"><span class="Apple-tab-span">	</span></span><span class="s5">b. </span><span class="s1">Additional columns may be included.</span></p>
<p class="p2"><span class="s1">Output Files:</span></p>
<p class="p4"><span class="s1">1.</span><span class="s3">     </span><span class="s1"><b>Tximport RDS file</b> An RDS file named “tximport.rds” containing data output by RSEM and re-formatted into a list of 3 matrices. The list contains 1) matrix with expected counts (“counts”), 2) matrix with expected length (“length”), and 3) matrix with transcripts per million (TPM) (“abundance”). For each matrix, genes are listed as rows and samples as columns.</span></p>
<p class="p2"><span class="s1"> </span></p>
<p class="p2"><span class="s2">Step #2:</span><span class="s1">  Normalize expression.</span></p>
<p class="p2"><span class="s1">Example Usage:</span></p>
<p class="p6"><span class="s1">Rscript normalizeExpression.R &lt;TXIMPORT RDS FILE&gt; &lt;GENE FILE&gt; &lt;SAMPLE FILE&gt; &lt;NORMALIZATION METHOD&gt; &lt;CORRECTION METHOD&gt; &lt;NEGATIVE CONTROL GENE FILE&gt;</span></p>
<p class="p2"><span class="s1">Input Parameters:</span></p>
<p class="p3"><span class="s1">1.</span><span class="s3">     </span><span class="s1"><b>Tximport RDS file </b>The tximport RDS file output from step #1.</span></p>
<p class="p3"><span class="s1">2.</span><span class="s3">     </span><span class="s1"><b>Gene file</b> A tab-delimited file containing the list of gene identifiers to include in the normalization and downstream steps.</span></p>
<p class="p9"><span class="s1"><span class="Apple-tab-span">	</span>a.</span><span class="s3">     </span><span class="s1">The file must minimally include the column(s) below.</span></p>
<p class="p10"><span class="s3"><span class="Apple-tab-span">	</span><span class="Apple-tab-span">	</span> </span><span class="s1">i.</span><span class="s3">     </span><span class="s1">“gene_id_simplified”: Ensembl gene identifier. The gene version number should be removed, e.g. “ENSG00000196230” instead of “ENSG00000196230.12”.</span></p>
<p class="p3"><span class="s1">3.</span><span class="s3">     </span><span class="s1"><b>Sample file</b> A tab-delimited file listing samples to use for RARE-Seq analysis. All samples included in this file must have also been included in the sample file used for step #1, and the same sample file can be used for both steps. </span></p>
<p class="p9"><span class="s1"><span class="Apple-tab-span">	</span>a.</span><span class="s3">     </span><span class="s1">The file must minimally include the column(s) below.</span></p>
<p class="p10"><span class="s3"><span class="Apple-tab-span">	</span><span class="Apple-tab-span">	</span> </span><span class="s1">i.</span><span class="s3">     </span><span class="s1">“ID”: Unique sample identifier</span></p>
<p class="p10"><span class="s3"><span class="Apple-tab-span">	</span><span class="Apple-tab-span">	</span> </span><span class="s1">ii.</span><span class="s3">     </span><span class="s1">“group”: Relevant sample categorization (i.e., “cancer” and “control”)</span></p>
<p class="p10"><span class="s3"><span class="Apple-tab-span">	</span><span class="Apple-tab-span">	</span> </span><span class="s1">iii.</span><span class="s3">     </span><span class="s1">“Reference”: Meta-reference samples may be marked “yes” or “NA” to indicate whether the sample is or is not included in the meta-reference subset, respectively.</span></p>
<p class="p9"><span class="s1"><span class="Apple-tab-span">	</span>b.</span><span class="s3">     </span><span class="s1">The file may include the columns below if step #7 is to be run.</span></p>
<p class="p10"><span class="s3"><span class="Apple-tab-span">	</span><span class="Apple-tab-span">	</span> </span><span class="s1">i.</span><span class="s3">     </span><span class="s1">“EN: Sample labels including “train”, “validation”, or “NA”, if they are to be used for model training, used for model validation, or excluded (i.e., meta-reference samples), respectively.</span></p>
<p class="p9"><span class="s1"><span class="Apple-tab-span">	</span>c.</span><span class="s3">     </span><span class="s1">Additional columns may be included.</span></p>
<p class="p3"><span class="s1">4.</span><span class="s3">     </span><span class="s1"><b>Normalization method</b> A character string for the appropriate normalization method and must be set to “NONE”, “TMM”, or “TPM”.</span></p>
<p class="p3"><span class="s1">5.</span><span class="s3">     </span><span class="s1"><b>Correction method</b> A character string for the appropriate correction method and must be set to “NONE”, “RUVg”, or “RUVs”.</span></p>
<p class="p3"><span class="s1">6.</span><span class="s3">     </span><span class="s1"><b>Negative control gene file</b> A tab-delimited file listing the genes to use for RUVg correction, and therefore is required if “RUVg” is specified above. For cfRNA analysis, the default is to use platelet marker genes for RUVg correction.</span></p>
<p class="p9"><span class="s1"><span class="Apple-tab-span">	</span>a.</span><span class="s3">     </span><span class="s1">The file must minimally include the column(s) below.</span></p>
<p class="p11"><span class="s3"><span class="Apple-tab-span">	</span><span class="Apple-tab-span">	</span> </span><span class="s1">i.</span><span class="s3">     </span><span class="s1">“gene_id_simplified”: Ensembl gene identifier. The gene version number should be removed, e.g. “ENSG00000196230” instead of “ENSG00000196230.12”.</span></p>
<p class="p2"><span class="s1">Output Files:</span></p>
<p class="p4"><span class="s1">2.</span><span class="s3">     </span><span class="s1"><b>Normalized RDS file</b> An RDS file named according to the &lt;NORMALIZATION METHOD&gt; and &lt;CORRECTION METHOD&gt;” input. The list contains 1) data.frame with input gene information (“genes”), 2) data.frame with input sample information (“samples”), 3) matrix of RSEM expected counts (“counts”), 4) matrix of RSEM expected length (“length”), and 5) matrix of normalized gene expression (“log2NX”). For last 3 matrices, genes are listed as rows and samples as columns.</span></p>
<p class="p2"><span class="s1"> </span></p>
<p class="p2"><span class="s2">Step #3:</span><span class="s1"> Calculate z-scores.</span></p>
<p class="p2"><span class="s1">Example Usage:</span></p>
<p class="p6"><span class="s1">Rscript calcZcores.R &lt;NORMALIZED RDS FILE&gt;</span></p>
<p class="p2"><span class="s1">Input Parameters:</span></p>
<p class="p4"><span class="s1">1.</span><span class="s3">     </span><span class="s1"><b>Normalized RDS file</b> The RDS file output from step #2.</span></p>
<p class="p2"><span class="s1">Output Files:</span></p>
<p class="p4"><span class="s1">1.</span><span class="s3">     </span><span class="s1"><b>Normalized RDS file with Z-scores</b> An RDS file containing the same matrices as output for step #2 but with the matrix of z-scores added and with <i>“_withZscores</i>” appended to the file name. If meta-reference samples were specified in the sample information file in step #2, these samples will be used for Z-score standardization and will therefore be removed from the Z-score matrix.</span></p>
<p class="p2"><span class="s1"><b> </b></span></p>
<p class="p2"><span class="s2">Step #4:</span><span class="s1"> Calculate enrichment scores.</span></p>
<p class="p2"><span class="s1">Example Usage:</span></p>
<p class="p6"><span class="s1">Rscript calcEnrichment.R &lt;NORMALIZED RDS FILE WITH ZSCORES&gt; &lt;ENRICHMENT METHOD&gt; &lt;GENE SIGNATURE FILE&gt; &lt;OUTPUT DIRECTORY&gt;</span></p>
<p class="p2"><span class="s1">Input Parameters:</span></p>
<p class="p3"><span class="s1">1.</span><span class="s3">     </span><span class="s1"><b>Normalized RDS file with Z-scores</b> The RDS file output from step #3.</span></p>
<p class="p3"><span class="s1">2.</span><span class="s3">     </span><span class="s1"><b>Enrichment method</b> A character string specifying the desired method for calculating enrichment scores. The default method is “<i>stouffer</i>”, which uses Stouffer’s method to combine multiple weighted Z-scores into a single gene signature enrichment score. The method may also be “<i>percentile_X</i>” and if so, X should be replaced by a numeric value in [0,1] (e.g., 0.9) to be used as the cut point among weighted Z-scores within the gene signature defining the gene signature enrichment score. Finally, the method may be set to “<i>sum</i>”, which simply adds together the weighted Z-scores to calculate the gene signature enrichment score.</span></p>
<p class="p3"><span class="s1">3.</span><span class="s3">     </span><span class="s1"><b>Gene signature(s) file</b> A tab-delimited file containing the list of genes in the gene signature(s) of interest.</span></p>
<p class="p9"><span class="s1">a.</span><span class="s3">     </span><span class="s1">The file must minimally include the column(s) below.</span></p>
<p class="p10"><span class="s3">                                               </span><span class="s1">i.</span><span class="s3">     </span><span class="s1">“gene_id_simplified”: Ensembl gene identifier. The gene version number should be removed, e.g. “ENSG00000196230” instead of “ENSG00000196230.12”.</span></p>
<p class="p9"><span class="s1">b.</span><span class="s3">     </span><span class="s1">The file may include the columns below.</span></p>
<p class="p10"><span class="s3">                                               </span><span class="s1">i.</span><span class="s3">     </span><span class="s1">“gene_set_name”: Gene set name(s). Multiple gene signatures may be run but must be distinguished with unique names in this column.</span></p>
<p class="p10"><span class="s3">                                              </span><span class="s1">ii.</span><span class="s3">     </span><span class="s1">“weight”: Gene weight(s) to use for the specified enrichment method. If no weights are provided, all gene weights are set to 1.</span></p>
<p class="p4"><span class="s1">4.</span><span class="s3">     </span><span class="s1"><b>Output directory</b> A character string providing the name of the directory in which to output the enrichment score results. Since this step can be run multiple times for multiple gene signature files, each run should use a unique output directory name to avoid overwriting data.</span></p>
<p class="p2"><span class="s1">Output Files:</span></p>
<p class="p4"><span class="s1">1.</span><span class="s3">     </span><span class="s1"><b>Enrichment scores file</b> A tab-delimited file containing the enrichment scores and associated empirical p-values for each sample for each gene signature.</span></p>
<p class="p2"><span class="s1"> </span></p>
<p class="p2"><span class="s2">Step #5:</span><span class="s1"> Identify differentially expressed genes.</span></p>
<p class="p2"><span class="s1">Example Usage:</span></p>
<p class="p6"><span class="s1">Rscript differentialExpression.R &lt;NORMALIZED RDS FILE&gt; &lt;DIFFERENTIAL EXPRESSION METHOD&gt; &lt;MODEL EQUATION&gt;</span></p>
<p class="p2"><span class="s1">Input Parameters:</span></p>
<p class="p3"><span class="s1">1.</span><span class="s3">     </span><span class="s1"><b>Normalized RDS file</b> The RDS file output from step #2.</span></p>
<p class="p3"><span class="s1">2.</span><span class="s3">     </span><span class="s1"><b>Differential expression method</b> A character string specifying which differential expression method to use and must be either “DESeq2” or “DESeq2shrink”. The latter runs log fold change (LFC) shrinkage using the ashr method to provide better gene ranking by effect size. More information is provided in the DESeq2 tutorial and vignettes (https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html).</span></p>
<p class="p4"><span class="s1">3.</span><span class="s3">     </span><span class="s1"><b>Model equation</b> A character string providing the equation used to model differential expression. This equation must use column headers that are present in the sample information matrix in the RDS file input (which come from the sample information file input in step #2). For example, for a simple case-control analysis, the model equation may be “~group” to determine differentially expressed genes between two (or more) sample groups as distinguished in this column. More information is provided in the DESeq2 tutorial and vignettes (https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html).</span></p>
<p class="p2"><span class="s1">Output Files:</span></p>
<p class="p4"><span class="s1">1.</span><span class="s3">     </span><span class="s1"><b>Differential expression file(s)</b> A tab-delimited file containing the log fold change and adjusted p-values for differential expression based on the provided model equation. One file will be output for each contrast evaluated.</span></p>
<p class="p2"><span class="s1"> </span></p>
<p class="p2"><span class="s2">Step #6:</span><span class="s1"> Set up elastic net training inputs.</span></p>
<p class="p2"><span class="s1">Example Usage:</span></p>
<p class="p6"><span class="s1">Rscript setupEN.R &lt;NORMALIZED RDS FILE&gt; &lt;GENOTYPING FILE&gt;</span></p>
<p class="p2"><span class="s1">Input Parameters:</span></p>
<p class="p3"><span class="s1">1.</span><span class="s3">     </span><span class="s1"><b>Normalized RDS file </b>The RDS file output from step #2. The sample info matrix must include a column using the header “EN” that specifies the subset of samples to be used for model training (“train”), the subset of samples to be used for model validation (“validation”), and meta-reference to be excluded from this step (“NA”).</span></p>
<p class="p4"><span class="s1">2.</span><span class="s3">     </span><span class="s1"><b>Genotyping file </b>(Optional)<b> </b>The tab-delimited file output by the targeted-caller script, named “variant_call_table.txt”. If not provided, only gene expression features will be used for model training.</span></p>
<p class="p2"><span class="s1">Output Files:</span></p>
<p class="p4"><span class="s1">3.</span><span class="s3">     </span><span class="s1"><b>Normalized RDS file for Caret</b> An RDS file with “<i>_forCaret.rds</i>” appended to the filename. The RDS object contains 1) data.frame of gene information that has been filtered to include only features to be used for model training, 2) data.frame of sample information that has been filtered to include only samples to be used for model training, 3) matrices of standardized features to be used as predictors for model training and validation, 4) binarized vectors to be used as outcome variable (e.g. case/1 or control/0) for model training and validation. See the <b>Description</b> section for more information about how feature filtering and pre-processing.</span></p>
<p class="p2"><span class="s1"> </span></p>
<p class="p2"><span class="s2">Step #7:</span><span class="s1"> Train elastic net model.</span></p>
<p class="p2"><span class="s1">Example Usage:</span></p>
<p class="p6"><span class="s1">Rscript trainEN.R &lt;NORMALIZED RDS FILE FOR CARET&gt;</span></p>
<p class="p2"><span class="s1">Input Parameters:</span></p>
<p class="p4"><span class="s1">1.</span><span class="s3">     </span><span class="s1"><b>Normalized RDS file for Caret</b> The RDS file output from step #7.</span></p>
<p class="p2"><span class="s1">Output Files:</span></p>
<p class="p3"><span class="s1">4.</span><span class="s3">     </span><span class="s1"><b>Model selected features file</b> A tab-delimited file named “model_selected_features.txt” that contains feature coefficients. A feature coefficient above zero indicates feature selection.</span></p>
<p class="p3"><span class="s1">5.</span><span class="s3">     </span><span class="s1"><b>Model sample predictions file</b> A tab-delimited file named “model_sample_predictions.txt” that contains the cancer likelihood score for all training samples.</span></p>
<p class="p3"><span class="s1">6.</span><span class="s3">     </span><span class="s1"><b>Model metrics file</b> A tab-delimited file named “model_metrics.txt” that contains area under the curve (AUC) calculated either using the full training set (“Train”) or using the 10-fold cross validation results (“CV”). </span></p>
<p class="p4"><span class="s1">7.</span><span class="s3">     </span><span class="s1"><b>Model RDS file </b>An RDS file named “model.rds” that contains the final model as output by the caret R package.</span></p>
<p class="p2"><span class="s1"> </span></p>
<p class="p2"><span class="s2">Step #8:</span><span class="s1"> Validate elastic net model.</span></p>
<p class="p2"><span class="s1">Example Usage:</span></p>
<p class="p6"><span class="s1">Rscript validateEN.R &lt;NORMALIZED RDS FILE FOR CARET&gt; &lt;MODEL RDS FILE&gt;</span></p>
<p class="p2"><span class="s1">Input Parameters:</span></p>
<p class="p3"><span class="s1">1.</span><span class="s3">     </span><span class="s1"><b>Normalized RDS file for Caret</b> The RDS file output from step #7.</span></p>
<p class="p4"><span class="s1">2.</span><span class="s3">     </span><span class="s1"><b>Model RDS file </b>The RDS file output from step #8.</span></p>
<p class="p2"><span class="s1">Output Files:</span></p>
<p class="p3"><span class="s1">1.</span><span class="s3">     </span><span class="s1"><b>Model sample predictions file</b> A tab-delimited file named “model_sample_predictions_validation.txt” that contains the cancer likelihood score for all withheld validation samples.</span></p>
<p class="p4"><span class="s1">2.</span><span class="s3">     </span><span class="s1"><b>Model metrics file </b>A tab-delimited file named “model_metrics_validation.txt” that contains the area under the curve *AC) calculated using the validation set (“Validation”).</span></p>
</body>
</html>
